examples/dpsgd_example.py

```diff
@@ -0,0 +1,185 @@
+# limitations under the License.
+
+"""JAX efficiently trains a differentially private conv net on MNIST.
+This script contains a JAX implementation of Differentially Private Stochastic
+Gradient Descent (https://arxiv.org/abs/1607.00133). DPSGD requires clipping
+the per-example parameter gradients, which is non-trivial to implement
+efficiently for convolutional neural networks.  The JAX XLA compiler shines in
+this setting by optimizing the minibatch-vectorized computation for
+convolutional architectures. Train time takes a few seconds per epoch on a
+commodity GPU.
+This code depends on tensorflow_privacy (https://github.com/tensorflow/privacy)
+  Install instructions:
+    $ pip install tensorflow git+https://github.com/tensorflow/privacy.git
+The results match those in the reference TensorFlow baseline implementation:
+  https://github.com/tensorflow/privacy/tree/master/tutorials
+Example invocations:
+  # this non-private baseline should get ~99% acc
+  python -m examples.differentially_private_sgd \
+    --dpsgd=False \
+    --learning_rate=.1 \
+    --epochs=20 \
+   this private baseline should get ~95% acc
+  python -m examples.differentially_private_sgd \
+   --dpsgd=True \
+   --noise_multiplier=1.3 \
+   --l2_norm_clip=1.5 \
+   --epochs=15 \
+   --learning_rate=.25 \
+  # this private baseline should get ~96.6% acc
+  python -m examples.differentially_private_sgd \
+   --dpsgd=True \
+   --noise_multiplier=1.1 \
+   --l2_norm_clip=1.0 \
+   --epochs=60 \
+   --learning_rate=.15 \
+  # this private baseline should get ~97% acc
+  python -m examples.differentially_private_sgd \
+   --dpsgd=True \
+   --noise_multiplier=0.7 \
+   --l2_norm_clip=1.5 \
+   --epochs=45 \
+   --learning_rate=.25 \
+"""
+
+import time
+import warnings
+
+from absl import app
+from absl import flags
+
+import jax
+import jax.numpy as jnp
+from jax.experimental import stax
+import optax
+
+import datasets
+
+# https://github.com/tensorflow/privacy
+from tensorflow_privacy.privacy.analysis.rdp_accountant import compute_rdp
+from tensorflow_privacy.privacy.analysis.rdp_accountant import get_privacy_spent
+
+NUM_EXAMPLES = 60_000
+FLAGS = flags.FLAGS
+
+flags.DEFINE_boolean(
+    'dpsgd', True, 'If True, train with DP-SGD. If False, '
+    'train with vanilla SGD.')
+flags.DEFINE_float('learning_rate', .15, 'Learning rate for training')
+flags.DEFINE_float('noise_multiplier', 1.1,
+                   'Ratio of the standard deviation to the clipping norm')
+flags.DEFINE_float('l2_norm_clip', 1.0, 'Clipping norm')
+flags.DEFINE_integer('batch_size', 256, 'Batch size')
+flags.DEFINE_integer('epochs', 60, 'Number of epochs')
+flags.DEFINE_integer('seed', 0, 'Seed for jax PRNG')
+flags.DEFINE_integer(
+    'microbatches', None, 'Number of microbatches '
+    '(must evenly divide batch_size)')
+flags.DEFINE_string('model_dir', None, 'Model directory')
+flags.DEFINE_float('delta', 1e-5, 'Target delta used to compute privacy spent.')
```

Comment:
NOTE: Not `delta` flag not present in original JAX example, added because it was easy to, could remove if we want to stay closer to the original.